### PR TITLE
[clang][analyzer] Add "pedantic" mode to StreamChecker.

### PR DESCRIPTION
**Pull request by balazske**
The checker may create failure branches for all stream write operations only if the new option "pedantic" is set to true.
Result of the write operations is often not checked in typical code. If failure branches are created the checker will warn for unchecked write operations and generate a lot of "false positives" (these are valid warnings but the programmer does not care about this problem).

### CODE REVIEW

File: clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp [hunk 1]
```diff
@@ -1059,6 +1066,9 @@ void StreamChecker::evalFputx(const FnDescription *Desc, const CallEvent &Call,
 
   // Add transition for the failed state. The resulting value of the file
   // position indicator for the stream is indeterminate.
+  if (!PedanticMode)
+    return;
+
```
Comment (NagyDonat) [inline clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:1071]:
```suggestion
  if (!PedanticMode)
    return;

  // Add transition for the failed state. The resulting value of the file
  // position indicator for the stream is indeterminate.
```

File: clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp [hunk 2]
```diff
@@ -1094,6 +1104,9 @@ void StreamChecker::evalFprintf(const FnDescription *Desc,
...
```
Comment (NagyDonat) [inline clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:1109]:
```suggestion
  if (!PedanticMode)
    return;

  // Add transition for the failed state. The resulting value of the file
  // position indicator for the stream is indeterminate.
```

File: clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp [hunk 3]
```diff
@@ -1264,16 +1277,18 @@ void StreamChecker::evalFseek(const FnDescription *Desc, const CallEvent &Call,
...
+  // Add success state.
   ProgramStateRef StateNotFailed = E.bindReturnValue(State, C, 0);
-  ProgramStateRef StateFailed = E.bindReturnValue(State, C, -1);
-
   // No failure: Reset the state to opened with no error.
   StateNotFailed =
       E.setStreamState(StateNotFailed, StreamState::getOpened(Desc));
   C.addTransition(StateNotFailed);
 
+  // Add failure state.
```
Comment (NagyDonat) [inline clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:1287]:
```suggestion
  // In pedantic mode, also add a failure state.
```

**Review by NagyDonat — APPROVED** [review]
Good change, I have a few minor suggestions that tweak comment placement.

I felt that it's confusing to have an "Add a transition for" comment directly followed by a seemingly unrelated `if (!PedanticMode)` block -- my suggestions swap the comment and the `if` block, but you could also e.g. prefix the comment with "In pedantic mode, also..." to clarify the logical connection between the `if` and the rest of the comment.

As an unrelated note, please explain this checker option to the documentation of the checker (in addition to the `Checkers.td` change).

**Review by steakhal — COMMENTED** [review]
Unless you plan to add more heuristics, I'd prefer a more concrete option name, like AssumeSuccessfulWrites=true.
This would better describe it imo.

**Comment by NagyDonat** [comment]
Personally I like the habit that these options are consistently named "Pedantic" because it highlights that they all provide "true, but not helpful" results. When the users see that an option is named "Pedantic", many of them will be able to immediately conclude that they are not interested; while if the option was named "AssumeSuccessfulWrites", they would need to understand its exact meaning (so they would probably read the full doc, not just the name) and its relevance in their project to make a decision.

I agree that "AssumeSuccessfulWrites" is more informative in a certain sense, but it still doesn't convey enough information for a full understanding; and "Pedantic" provides a different kind of information that (IMO) is more practical overall.

**Comment by balazske** [comment]
> Unless you plan to add more heuristics, I'd prefer a more concrete option name, like AssumeSuccessfulWrites=true. This would better describe it imo.

I do not like totally the name "Pedantic", it could be "AssumeOftenUncheckedOperationsMayFail". I am not sure if this behavior is needed only on write operations, the intent was to remove failure branches from all operations that are often unchecked.

**Comment by NagyDonat** [comment]
I feel that "AssumeOftenUncheckedOperationsMayFail" does not provide more information than "Pedantic" (≈ report issues that are often left unchecked), while it is significantly longer, so my preferences are _Pedantic > AssumeSuccessfulWrites > AssumeOftenUncheckedOperationsMayFail_.

**Comment by steakhal** [comment]
> > Unless you plan to add more heuristics, I'd prefer a more concrete option name, like AssumeSuccessfulWrites=true. This would better describe it imo.
> 
> I do not like totally the name "Pedantic", it could be "AssumeOftenUncheckedOperationsMayFail". I am not sure if this behavior is needed only on write operations, the intent was to remove failure branches from all operations that are often unchecked.

The docs within `checkers.td` says `If false, assume that stream write operations do never fail.`. Given this and your last reply, maybe we should consider elaborating the description to reflect that more operations could be affected by setting this option.

> I feel that "AssumeOftenUncheckedOperationsMayFail" does not provide more information than "Pedantic" (≈ report issues that are often left unchecked), while it is significantly longer, so my preferences are _Pedantic > AssumeSuccessfulWrites > AssumeOftenUncheckedOperationsMayFail_.

I think we can disagree.
Given that this isn't such a serious change, it should be fine to continue as you planned.

One remark, inside the `clang/test/Analysis/stream-pedantic.c` could we have `no-warning` markers for the places where the pedantic mode would raise issues?

File: clang/docs/analyzer/checkers.rst [hunk 4]
```diff
@@ -3138,10 +3138,13 @@ are detected:
...
-and ``FILE *`` pointers. Operations on standard streams like ``stdin`` are not
-treated specially and are therefore often not recognized (because these streams
-are usually not opened explicitly by the program, and are global variables).
+The checker assumes by default that any of the stream operations, except the
+write operations (like ``fwrite``, ``fprintf``) can fail. The option
+``Pedantic`` controls if the write operations are assumed to be successful or
+not. If set to ``true``, the write operations (including change of file
+position) can fail too (default is ``false``). This option was added to prevent
+many checker warnings if the code contains many write operations without error
+check.
```
Comment (NagyDonat) [inline clang/docs/analyzer/checkers.rst:3147]:
```suggestion
When this checker models a stream operation, it usually splits the analysis into
two cases: one where the stream operation succeeds, and one where it fails.
However, in the case of write operations (like ``fwrite``, ``fprintf`
and even `fsetpos`) this behavior could produce a large amount of unwanted
reports on projects that don't have error checks around the write operations --
so by default the checker assumes that write operations always succeed.

This behavior can be controlled by the ``Pedantic`` flag: with
``-analyzer-config alpha.unix.Stream:Pedantic=true`` the checker will model the
cases where a write operation fails and report situations where this leads to
erroneous behavior. (The default is ``Pedantic=false``, where write operations
are assumed to succeed.)
```
I felt that this paragraph was a bit difficult to understand, so I rewrote it in a somewhat more verbose and hopefully clearer way. Feel free to tweak this if you feel that there could be further improvements.

**Review by NagyDonat — APPROVED** [review]
Overall LGTM, thanks for improving this checker!

I reworded the description of the `Pedantic` mode in the documentation because I felt that it's difficult to understand. (It's difficult to explain this topic clearly...)

**Review by NagyDonat — APPROVED** [review]
Thanks for the update! Feel free to merge the commit and initiate the de-alpha process.

**Pull request merged**